design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,137 @@
+# ObjectBucketClaim CRD and Controller
+
+## Proposed Feature
+
+**A generalized method of S3 bucket provisioning through the implementation of an `ObjectBucketClaim` CRD and controller.** 
+
+Users will request buckets via the Kubernetes API and have returned ConfigMaps with connection information for that bucket.  
+
+One ConfigMap will be created per `ObjectBucketClaim` instance.  ConfigMap names will be deterministic and derived from the `ObjectBucketClaim` name.  This gives users the opportunity to define their Pod spec at the same time as the `ObjectBucketClaim`.  Additionally, the `ObjectBucketClaim` and Pod can be deployed at the same time thanks in-built synchronization in Kubernetes.
+
+Bucket deletion in the early stages of this controller design will not be addressed beyond cleaning up API objects.  Deletion of the bucket within the object store is left as an administrator responsibility.  This is to prevent the accidental destruction of data by users.
+
+## Design
+
+### Work Flow
+
+#### Assumptions
+
+- An object store has been deployed in a Kubernetes Cluster ([e.g. Rook-Ceph Object](https://rook.github.io/docs/rook/master/ceph-object.html))
+- A `Secret` (defined below) containing the Access Key and Secret Key has been created.
+- A `StorageClass` (defined below) has been created for the object store service.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want leverage the Kubernetes API to create S3 buckets. I expect to get back the bucket connection information in a ConfigMap._
+ 
+![work flow](./ObjectBucketClaim.png)
+
+1. The Rook Operator detects a new `ObjectBucketClaim` instance.  
```

Comment:
>which operator is watching for new ObjectBucketClaim instances to be created? 
The claim controller will run inside it's own operator the same as the other rook operators (ceph, nfs, edgefs, etc).

> Does that operator/controller have the general responsibility of provisioning buckets for any s3 compatible object store?
yes.  the operator should be agnostic to the store provider.  the only requirement is that the provider exposes an S3 interface

Reply:
> If we want to provision minio or s3 buckets we'd set the StorageClass provisioner set to some other provisioner (rook minio, or something that does s3, or whatever).

The implication here being bucket provisioning would be implemented redundantly across each object store providers (in their operators).  The convention for object stores to expose an S3 interface means that we only need 1 operator which implements and S3 API client.  It can be completely agnostic to the actual storage provider.

Reply:
> wouldn't each provisioner only understand one particular object store type (ceph in rook's case, at least initially)?

As far as the object _store_, yes since each store provider has its own requirements for provisioning (CLI vs API calls, etc).  However, buckets creation is abstracted thru the S3 interface.  A single bucket operator can remain unaware of the underlying store type so long as they expose and S3 endpoint.